### PR TITLE
[FBref] Handle missing Scores & Fixtures page on the Big 5 European Leagues Stats page

### PR DESCRIPTION
**Pull request by lorenzodb1**
- For every n rows, the website adds a row in a table that replicates the table header. This caused read_schedule to fail as the number of rows in df_table would be higher than the one of the list of match URLs obtained (see
https://github.com/probberechts/soccerdata/issues/277). I added the logic to remove those replicated headers when found.
- The website has no specific Scores & Fixtures on the Big 5 European Leagues Stats page. Thus it'd go to the generic Scores & Fixtures page, which shows games currently being played. Because of this, I had to move the optimisation that combines the top five leagues under that label in read_leagues, as read_schedule necessarily needs the five top leagues separately rather than in their combined form.

### CODE REVIEW

**Comment by lorenzodb1** [comment]
Tests seem to be failing due to the issue fixed in https://github.com/probberechts/soccerdata/pull/286

**Comment by lorenzodb1** [comment]
@probberechts is there a way we can get this and #286 merged and released any time soon? I'd really appreciate that as those bugs are blocking me in my personal project :)

**Comment by probberechts** [comment]
Hi @lorenzodb1 

I'll have to take a closer look at these and I'll try to make some time during the weekend. What's blocking me from merging them immediately is:

- #284 complicates the API of the `read_leagues` and `read_seasons` methods. I would like to check whether there isn't another way to implement this that does not require adding these vague parameters to public methods.
- The example given in #286 refers to an unsupported league. If it is something specific to that league, I might not even fix it. 
- I am worried that the changes made in #286 would break other things. It might not throw an error anymore, but it looks like it will no longer be able to merge columns anymore in some cases.

**Comment by probberechts** [comment]
For dealing with the missing Scores & Fixtures on the Big 5 European Leagues Stats page, I would be ok with two alternative approaches:
- raise an exception when `read_schedule` is called for the 'Big 5 European Leagues Combined',  suggesting creating a new instance of the FBref scraper with `leagues = ['ENG-Premier League', 'FRA-Ligue 1', ...]`
- follow your approach of modifying the `read_leagues` method, but I would add a more informative "split_up_big5 = False" parameter to both `read_leagues` and `read_seasons`

**Comment by lorenzodb1** [comment]
> follow your approach of modifying the `read_leagues` method, but I would add a more informative "split_up_big5 = False" parameter to both `read_leagues` and `read_seasons`

This was my first approach to fixing that issue. I changed it to the current implementation because it would mean adding the same logic of splitting the Big5 leagues in both methods. Since `read_leagues` is only called by `read_seasons`, it could make sense to change the signature to just passing the leagues that need to be considered. If you have strong opinions against that, though, I'll be more than happy to change the approach to what you described.

Out of curiosity, what's the reason behind keeping both `read_leagues` and `read_seasons` methods public?

**Comment by probberechts** [comment]
Unless I am missing something, the only thing that would have to be changed in `read_seasons`  are [L162](https://github.com/probberechts/soccerdata/blob/916e3e83219bbe021ff1f69efb5b1f2e7fdf6290/soccerdata/fbref.py#L162C9-L162C41) to `df_leagues = self.read_leagues(split_up_big_5)` and drop the `self.leagues` filter in [L196](https://github.com/probberechts/soccerdata/blob/916e3e83219bbe021ff1f69efb5b1f2e7fdf6290/soccerdata/fbref.py#L196).

> Out of curiosity, what's the reason behind keeping both read_leagues and read_seasons methods public?

These contain relevant data as well, such as the season's champion and top scorer. Additionaly, making them private would be a breaking change.

**Comment by lorenzodb1** [comment]
> Unless I am missing something, the only thing that would have to be changed in `read_seasons` are [L162](https://github.com/probberechts/soccerdata/blob/916e3e83219bbe021ff1f69efb5b1f2e7fdf6290/soccerdata/fbref.py#L162C9-L162C41) to `df_leagues = self.read_leagues(split_up_big_5)` and drop the `self.leagues` filter in [L196](https://github.com/probberechts/soccerdata/blob/916e3e83219bbe021ff1f69efb5b1f2e7fdf6290/soccerdata/fbref.py#L196).

If the filter in [L196](https://github.com/probberechts/soccerdata/blob/916e3e83219bbe021ff1f69efb5b1f2e7fdf6290/soccerdata/fbref.py#L196) can be dropped then that'd definitely be a better solution.

> These contain relevant data as well, such as the season's champion and top scorer. Additionaly, making them private would be a breaking change.

I see. Maybe t'd be worth adding them to the API docs then.

**Comment by probberechts** [comment]
> I see. Maybe it'd be worth adding them to the API docs then.

Yes, that makes sense. I've added them in #307.

**Comment by lorenzodb1** [comment]
@probberechts I applied your suggestion to these changes. I couldn't drop the `self.leagues` filter in [L196](https://github.com/probberechts/soccerdata/blob/916e3e83219bbe021ff1f69efb5b1f2e7fdf6290/soccerdata/fbref.py#L196) as it'd make all tests fail, but I changed it so I could still stop using `self.leagues` in it.

File: soccerdata/fbref.py [hunk 1]
```diff
@@ -120,9 +120,15 @@ def _is_complete(self, league: str, season: str) -> bool:
...
 
-    def read_leagues(self) -> pd.DataFrame:
+    def read_leagues(self, optimise_big5: bool = True) -> pd.DataFrame:
         """Retrieve selected leagues from the datasource.
 
+        Parameters
+        ----------
+        optimise_big5: bool
+            If True, it will load the "Big 5 European Leagues Combined" instead of
+            each league individually.
```
Comment (probberechts) [inline soccerdata/fbref.py:130]: That's not what it does! Consider the example below. According to the docstring, it should return a data frame with a single row with league="Big 5 European Leagues Combined". Please change to `split_up_big5 = False` as I suggested or change the implementation accordingly.

```py
fbref = sd.FBref(["ENG-Premier League", "FRA-Ligue 1", "ITA-Serie A", "ESP-La Liga", "GER-Bundesliga"], "2223")
fbref.read_leagues(optimise_big5 = True)
```

File: soccerdata/fbref.py [hunk 2]
```diff
@@ -193,7 +216,8 @@ def read_seasons(self) -> pd.DataFrame:
         df.drop_duplicates(subset=["league", "season"], keep="first", inplace=True)
         df = df.set_index(["league", "season"]).sort_index()
         return df.loc[
-            df.index.isin(list(itertools.product(self.leagues, self.seasons))), ["format", "url"]
+            df.index.isin(list(itertools.product(df_leagues.index.unique(), self.seasons))),
+            ["format", "url"],
```
Comment (probberechts) [inline soccerdata/fbref.py:220]: This is equivalent to: `df.loc[(slice(None), self.seasons) , ["format", "url"]]`.

**Review by probberechts — CHANGES_REQUESTED** [review]

File: soccerdata/fbref.py (hunk 1 above)
Comment (lorenzodb1) [reply soccerdata/fbref.py:129]: Good catch! I changed the implementation accordingly.

**Comment by probberechts** [comment]
The current implementation would be a breaking change and it still does not work as I would expect. Now it would scrape the data from the "Big 5 European Leagues Combined" page by default.  The code below would combine the top-5 leagues, but it should return 5 leagues:

```py
fbref = sd.FBref(["ENG-Premier League", "FRA-Ligue 1", "ITA-Serie A", "ESP-La Liga", "GER-Bundesliga"], "2223")
fbref.read_leagues()
```

**Comment by lorenzodb1** [comment]
@probberechts I ended up changing it to `split_up_big5: bool = False` like you recommended. Hopefully, that will be enough to merge and release these fix. Let me know if you have any other issues with these changes.

**Review by probberechts — APPROVED** [review]

**Comment by probberechts** [comment]
Looks good now. Thanks for all your work on improving the FBref scraper!

**Pull request merged**